### PR TITLE
*:  Anonymous field for RegionInfo - part 3

### PR DESCRIPTION
**Pull request by nolouch**
## What have you changed? (required)

part3 for https://github.com/pingcap/pd/pull/1207
## What are the type of the changes (required)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## Refer to a related PR or issue link (optional)
https://github.com/pingcap/pd/pull/1205
https://github.com/pingcap/pd/pull/1207
https://github.com/pingcap/pd/pull/1208

### CODE REVIEW

File: server/core/region_tree_test.go [hunk 1]
```diff
@@ -67,16 +67,16 @@ func (s *testRegionSuite) TestRegionInfo(c *C) {
...
+	r.meta.Peers = append(r.meta.Peers, removePeer)
 	c.Assert(DiffRegionPeersInfo(info, r), Matches, "Add peer.*")
 	c.Assert(DiffRegionPeersInfo(r, info), Matches, "Remove peer.*")
 	c.Assert(r.GetStorePeer(n), DeepEquals, removePeer)
-	r.RemoveStorePeer(n)
+	r = r.Clone(WithRemoveStorePeer(n))
 	c.Assert(DiffRegionPeersInfo(r, info), Equals, "")
 	c.Assert(r.GetStorePeer(n), IsNil)
-	r.Region.StartKey = []byte{0}
+	r.meta.StartKey = []byte{0}
```
Comment (rleungx) [inline server/core/region_tree_test.go]: should we use `core.WithStartKey` here?

File: server/core/region_tree_test.go [hunk 2]
```diff
@@ -67,16 +67,16 @@ func (s *testRegionSuite) TestRegionInfo(c *C) {
 		Id:      n,
 		StoreId: n,
 	}
-	r.Peers = append(r.Peers, removePeer)
+	r.meta.Peers = append(r.meta.Peers, removePeer)
```
Comment (rleungx) [inline server/core/region_tree_test.go]: when do we use `meta` or`GetMeta()`?

Comment (nolouch) [reply server/core/region_tree_test.go]: the test is in the `core` package, I do not change its behavior in the test of this package.

File: server/api/admin_test.go [hunk 3]
```diff
@@ -47,18 +47,18 @@ func (s *testAdminSuite) TestDropRegion(c *C) {
 	cluster := s.svr.GetRaftCluster()
 
 	// Update region's epoch to (100, 100).
-	region := cluster.GetRegionInfoByKey([]byte("foo"))
-	region.RegionEpoch.ConfVer, region.RegionEpoch.Version = 100, 100
+	region := cluster.GetRegionInfoByKey([]byte("foo")).Clone()
+	region.GetRegionEpoch().ConfVer, region.GetRegionEpoch().Version = 100, 100
```
Comment (disksing) [inline server/api/admin_test.go]: Need to use `SetRegionEpoch`

File: server/api/admin_test.go [hunk 4]
```diff
@@ -47,18 +47,18 @@ func (s *testAdminSuite) TestDropRegion(c *C) {
...
 	err := cluster.HandleRegionHeartbeat(region)
 	c.Assert(err, IsNil)
 
 	// Region epoch cannot decrease.
-	region.RegionEpoch.ConfVer, region.RegionEpoch.Version = 50, 50
+	region.GetRegionEpoch().ConfVer, region.GetRegionEpoch().Version = 50, 50
```
Comment (disksing) [inline server/api/admin_test.go]: Need to use `SetEpoch`.

File: server/api/trend_test.go [hunk 5]
```diff
@@ -55,21 +55,21 @@ func (s *testTrendSuite) TestTrend(c *C) {
 	c.Assert(op, NotNil)
 	newPeerID := op.Step(0).(schedule.AddLearner).PeerID
 	region := s.newRegionInfo(5, "a", "b", 3, 2, []uint64{1, 2}, []uint64{3}, 1)
-	region.Learners[0].Id = newPeerID
+	region.GetLearners()[0].Id = newPeerID
```
Comment (disksing) [inline server/api/trend_test.go]: Need to use `SetPeers()`?

File: server/api/trend_test.go [hunk 6]
```diff
@@ -55,21 +55,21 @@ func (s *testTrendSuite) TestTrend(c *C) {
...
 	mustRegionHeartbeat(c, svr, region)
 	region = s.newRegionInfo(5, "a", "b", 4, 2, []uint64{1, 3}, nil, 1)
-	region.Peers[1].Id = newPeerID
+	region.GetPeers()[1].Id = newPeerID
```
Comment (disksing) [inline server/api/trend_test.go]: Need to use `SetPeers()`?

File: server/api/trend_test.go [hunk 7]
```diff
@@ -55,21 +55,21 @@ func (s *testTrendSuite) TestTrend(c *C) {
...
 	mustRegionHeartbeat(c, svr, region)
 
 	op, err = svr.GetHandler().GetOperator(6)
 	c.Assert(err, IsNil)
 	c.Assert(op, NotNil)
 	newPeerID = op.Step(0).(schedule.AddLearner).PeerID
 	region = s.newRegionInfo(6, "b", "", 3, 2, []uint64{1, 2}, []uint64{3}, 1)
-	region.Learners[0].Id = newPeerID
+	region.GetLearners()[0].Id = newPeerID
```
Comment (disksing) [inline server/api/trend_test.go]: Need to use `SetPeers()`?

File: server/api/trend_test.go [hunk 8]
```diff
@@ -55,21 +55,21 @@ func (s *testTrendSuite) TestTrend(c *C) {
...
 	mustRegionHeartbeat(c, svr, region)
 	region = s.newRegionInfo(6, "b", "", 4, 2, []uint64{2, 3}, nil, 2)
-	region.Peers[1].Id = newPeerID
+	region.GetPeers()[1].Id = newPeerID
```
Comment (disksing) [inline server/api/trend_test.go]: Need to use `SetPeers()`?

**Pull request merged**